Add collect-pgo-profiles script to support collection on iOS.

dewei-zhu opened the pull request:
#### 6137efe4895017b2236d516299e70acab922e54a
<pre>
Add collect-pgo-profiles script to support collection on iOS.
<a href="https://bugs.webkit.org/show_bug.cgi?id=290715">https://bugs.webkit.org/show_bug.cgi?id=290715</a>
<a href="https://rdar.apple.com/148177797">rdar://148177797</a>

Reviewed by Stephanie Lewis.

Move PGO profile collection to a standalone script from &apos;build-and-collect-pgo-profiles&apos;.
Move &apos;llvm-profdata&apos; operations from &apos;pgo-profile&apos; into a separate module and make it be
able to try different `llvm-profdata` binaries.

* Tools/Scripts/build-and-collect-pgo-profiles: Moved collect pgo profile logic into
collect-pgo-profiles script.
* Tools/Scripts/collect-pgo-profiles: This script supports passing additional arguments
to run-benchmark and specifying a custom run-benchmark script.
(check_or_create_empty_directory):
(collect_pgo_profiles):
(main):
* Tools/Scripts/pgo-profile: Move llvm-profdata operations to &apos;llvm_profile_utils&apos;.
(pad):
(shortened):
(assert_directory):
(assert_file):
(summarize_parser):
(summarize):
(merge_parser):
(merge):
(combine_parser):
(combine):
(compress):
(shell): Deleted.
* Tools/Scripts/webkitpy/llvm_profile_utils.py: Modules for llvm-profdata operations.
(locate_xcrun_binary):
(simplify_profile_weights):
(ExecutablesFromEnvAndXcode):
(ExecutablesFromEnvAndXcode.detect_binaries):
(ExecutablesFromEnvAndXcode.preference_ordered_paths):
(ExecutablesFromEnvAndXcode.run):
(LLVMProfDataExecutable):
(LLVMProfileData):
(LLVMProfileData.show):
(LLVMProfileData.merge):
(LLVMProfileData.compress):
(merge_raw_profiles_in_directory_by_prefixes):

Canonical link: <a href="https://commits.webkit.org/293342@main">https://commits.webkit.org/293342@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/651d0108fa69771ad8b5ec9e09df33750bc58bf5

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/98632 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/18265 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/8502 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/103756 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/49219 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/100676 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/18558 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/26716 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/75084 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/32232 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/101636 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/14081 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/89076 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/55441 "Found 1 new API test failure: /WPE/TestSSL:/webkit/WebKitWebView/ephemeral-tls-errors (failure)") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/98117 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/13862 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/7041 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/48602 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/83816 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/7119 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/106128 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/25722 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/18744 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/84056 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/26099 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/85277 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/83541 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21100 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/28185 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/5867 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/19414 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/25680 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/25498 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/28818 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/27073 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

gsnedders commented on Tools/Scripts/collect-pgo-profiles:
```diff
@@ -0,0 +1,91 @@
+PGO_PROFILE = os.path.join(SCRIPT_DIR, 'pgo-profile')
+
+
+def ensure_emtpy_directory(path):
```
```suggestion
def ensure_empty_directory(path):
```

gsnedders commented on Tools/Scripts/collect-pgo-profiles:
```diff
@@ -0,0 +1,91 @@
+    parser.add_argument('--run-benchmark-harness', '-r', required=False, default=DEFAULT_RUN_BENCHMARK,
+                        help=f'Location for run-benchmark harness, defaults to {DEFAULT_RUN_BENCHMARK}')
+    parser.add_argument('--benchmarks', '-b', nargs='+', required=True,
+                        help='Benchmarks to run for PGO profile collection.')
```
`nargs='+'` for an option leads to some kinda weird parsing.

Given we pass the unknown arguments to `run_benchmark_args`, and that [accepts arbitrary arguments](https://github.com/WebKit/WebKit/blob/0e38e60fd027984816b1077e5b7c20a72ae94750/Tools/Scripts/webkitpy/benchmark_runner/run_benchmark.py#L68), you could imagine something like `collect-pgo-profiles --benchmarks speedometer argument-for-browser` and that will result in `(Namespace(benchmarks=['speedometer', 'argument-for-browser'], test=False), [])`.

dewei-zhu replied:
This is a good point. But to my understand, run-benchmark does not take any positional arguments, which makes this particular case not a problem.

gsnedders commented on Tools/Scripts/collect-pgo-profiles:
```diff
@@ -0,0 +1,91 @@
+def main():
+    parser = argparse.ArgumentParser(prog='collect-pgo-profiles',
+                                     description='PGO profile collection script, '
+                                                 'unknown arguments will be passed to run-benchmark harness.')
```
We could also `from webkitpy.benchmark_runner.run_benchmark import config_argument_parser` and then `parser = argparse.ArgumentParser(..., parents=[config_argument_parser()]`, which would then make all those options appear in `--help`, though we would then need to re-serialize them to pass them onwards. Not sure if that's worthwhile doing, but it avoids some of the problems with `nargs`.

gsnedders commented on Tools/Scripts/pgo-profile:
```diff
@@ -143,23 +137,32 @@ def compress(args):
+if __name__ == '__main__':
     parser = argparse.ArgumentParser(prog='pgo-profile')
+    verbose_parser = argparse.ArgumentParser(add_help=False)
+    verbose_parser.add_argument('--verbose', '-v', action='store_true', default=False, help='Turn on debug logging')
```
Adding a separate parser here seems surprising. Why are we not just adding this to `parser`?

dewei-zhu replied:
Adding to `parser` would conflict with `--help` option, unless we want to make parser created with `add_help=False`

gsnedders replied:
But why do we need it to be a separate parser at all? I don't understand that. Why do we want this to be part of the subparsers?

dewei-zhu replied:
This is because we want to support `pgo-profile merge ... -v`.
If this is not part of subparser, it will have to be `pgo-profile -v merge ....`

An less optimal alternative is repeat `.add_argument('--verbose', '-v', ...` for every single subparser.

gsnedders commented on Tools/Scripts/webkitpy/llvm_profile_utils.py:
```diff
@@ -0,0 +1,148 @@
+
+    @classmethod
+    def compress(cls, input_profile, output_file):
+        return subprocess.run(['/usr/bin/compression_tool', '-encode', '-i', input_profile, '-o', output_file,
```
Are there reasons to be running `/usr/bin/compression_tool` and not `compression_tool`?

dewei-zhu replied:
Using full path is more secure in terms of security to avoid attack like path injection.

gsnedders replied:
Sure, just slightly surprised we're concerned about this here but not with any of the other commands we run.

dewei-zhu replied:
Ah, I missed `c++filt`, this should be fixed now.

gsnedders commented on Tools/Scripts/collect-pgo-profiles:
```diff
@@ -0,0 +1,92 @@
+    if args.verbose:
+        logger.setLevel(logging.DEBUG)
+
+    ensure_empty_directory(args.output_directory)
```
I do wonder if we should just fail if the output directory is non-empty, rather than implicitly deleting it? But I haven't too much of a stake in this.

sundiamonde replied:
or maybe just append a timestamp?

sundiamonde replied:
or even better revision/build info

dewei-zhu replied:
Adding timestamp would require a fuzzing matching if we want to get the actual output directory, which may have multiple entries with the same prefix.
I guess I can fail if target directory exists but not empty.

gsnedders commented on Tools/Scripts/collect-pgo-profiles:
```diff
@@ -0,0 +1,92 @@
+        combine_command.extend([f'--{benchmark}', profile_directory])
+
+    combined_profile_directory = os.path.join(args.output_directory, 'output')
+    ensure_empty_directory(combined_profile_directory)
```
Is there any reason not to just use `os.makedir` here? Given we started with an empty directory, it would be very surprising if it existed.

gsnedders commented on Tools/Scripts/collect-pgo-profiles:
```diff
@@ -0,0 +1,92 @@
+    subprocess.run(combine_command, check=True)
+
+    compressed_profile_directory = os.path.join(args.output_directory, args.compressed_profile_sub_path)
+    ensure_empty_directory(compressed_profile_directory)
```
Is there any reason not to just use `os.makedir` here? Given we started with an empty directory, it would be very surprising if it existed.

gsnedders commented on Tools/Scripts/webkitpy/llvm_profile_utils.py:
```diff
@@ -0,0 +1,148 @@
+    @cache
+    def detect_binaries(cls):
+        llvm_profdata_binaries = []
+        process = subprocess.run(['which', cls.EXECUTABLE_NAME], check=False, capture_output=True, text=True)
```
`shutil.which` is almost certainly gonna be a fair bit quicker than calling a subprocess here.

gsnedders commented on Tools/Scripts/webkitpy/llvm_profile_utils.py:
```diff
@@ -0,0 +1,148 @@
+logger = logging.getLogger(__name__)
+
+
+def locate_xcrun_binary(sdk, binary_name):
```
Maybe `which_in_sdk` or something? The name of the function makes it sound more like we're finding `xcrun` itself.

dewei-zhu replied:
`which_in_sdk` might be a bit misleading, I would go with `locate_binary_via_xcrun`

sundiamonde commented on Tools/Scripts/collect-pgo-profiles:
```diff
@@ -0,0 +1,92 @@
+    with tempfile.TemporaryDirectory() as temp_dir:
+        output_file = os.path.join(temp_dir, f'{benchmark}.json')
+        command = [sys.executable, harness, '--plan', benchmark, '--diagnose-directory', benchmark_profile_directory,
+                   '--generate-pgo-profiles', '--http-server-type', 'builtin', '--count', '1',
```
we want to use twisted don't we?  That is what we are using by default in the lab

dewei-zhu replied:
`Tools/Scripts/build-and-collect-pgo-profiles` was using builtin. I can try if the environment works with twisted.

dewei-zhu replied:
Move to twisted.

dewei-zhu committed:
Add collect-pgo-profiles script to support collection on iOS.
https://bugs.webkit.org/show_bug.cgi?id=290715
rdar://148177797

Reviewed by Stephanie Lewis.

Move PGO profile collection to a standalone script from 'build-and-collect-pgo-profiles'.
Move 'llvm-profdata' operations from 'pgo-profile' into a separate module and make it be
able to try different `llvm-profdata` binaries.

* Tools/Scripts/build-and-collect-pgo-profiles: Moved collect pgo profile logic into
collect-pgo-profiles script.
* Tools/Scripts/collect-pgo-profiles: This script supports passing additional arguments
to run-benchmark and specifying a custom run-benchmark script.
(check_or_create_empty_directory):
(collect_pgo_profiles):
(main):
* Tools/Scripts/pgo-profile: Move llvm-profdata operations to 'llvm_profile_utils'.
(pad):
(shortened):
(assert_directory):
(assert_file):
(summarize_parser):
(summarize):
(merge_parser):
(merge):
(combine_parser):
(combine):
(compress):
(shell): Deleted.
* Tools/Scripts/webkitpy/llvm_profile_utils.py: Modules for llvm-profdata operations.
(locate_xcrun_binary):
(simplify_profile_weights):
(ExecutablesFromEnvAndXcode):
(ExecutablesFromEnvAndXcode.detect_binaries):
(ExecutablesFromEnvAndXcode.preference_ordered_paths):
(ExecutablesFromEnvAndXcode.run):
(LLVMProfDataExecutable):
(LLVMProfileData):
(LLVMProfileData.show):
(LLVMProfileData.merge):
(LLVMProfileData.compress):
(merge_raw_profiles_in_directory_by_prefixes):

Canonical link: https://commits.webkit.org/293342@main
